File: Scripts/goto.py
# Go to a requested generation.  The given generation can be an
# absolute number like 1,000,000 (commas are optional) or a number
# relative to the current generation like +9 or -6.  If the target
# generation is less than the current generation then we go back
# to the starting generation (normally 0) and advance to the target.
# Authors: Andrew Trevorrow and Dave Greene, April 2006.
# Updated Sept-Oct 2006 -- XRLE support and reusable default value.

from glife import getstring, validint
from time import time
import os
import golly

# --------------------------------------------------------------------

def goto(gen):
   currgen = int(golly.getgen())
   if gen[0] == '+':
      n = int(gen[1:])
      newgen = currgen + n
   elif gen[0] == '-':
      n = int(gen[1:])
      if currgen > n:
         newgen = currgen - n
      else:
         newgen = 0
   else:
      newgen = int(gen)
   
   if newgen < currgen:
      # try to go back to starting gen (not necessarily 0) and
      # then forwards to newgen; note that reset() also restores
      # rule and hashing option, so too bad if user changed
      # those options after the starting info was saved;
      # first save current location and scale
      midx, midy = golly.getpos()
      mag = golly.getmag()
      golly.reset()
      # restore location and scale
      golly.setpos(midx, midy)
      golly.setmag(mag)
      # current gen might be > 0 if user loaded a pattern file
      # that set the gen count
      currgen = int(golly.getgen())
      if newgen < currgen:
         golly.error("Can't go back any further; pattern was saved "\
         + "at generation " + str(currgen) + ".")
         return
   
   golly.show("Hit escape to abort...")
   oldsecs = time()
   while currgen < newgen:
      if golly.empty():
         golly.show("Pattern is empty.")
         return
      # using golly.step() as a faster method for getting to newgen
      # is left as an exercise for the reader :)
      golly.run(1)
      currgen += 1
      golly.dokey( golly.getkey() )    # allow keyboard interaction
      newsecs = time()
      if newsecs - oldsecs >= 1.0:     # do an update every sec
         oldsecs = newsecs
         golly.update()
   golly.show("")

# --------------------------------------------------------------------
GotoINIFileName = os.path.join(golly.appdir() + "Scripts","goto.ini")
previousgen=prevdefault=""
if os.access(GotoINIFileName, os.R_OK):
   f=open(GotoINIFileName, 'r')
   previousgen=f.readline()
   f.close()
   # should maybe change validint in __init__.py
   # (add "if len(s)==0: return False")
   # to avoid this initial conditional...
   if len(previousgen):
      if not validint(previousgen):
         previousgen=""
if previousgen!="":
   prevdefault=" (default is " + previousgen + ")"

gen = getstring("Go to what generation" + prevdefault + "?")
if len(gen) == 0:
   gen = previousgen
if len(gen) == 0:
   golly.show("")
elif gen=="+" or gen=="-":
   previousgen="" # provides a way to clear the default
elif not validint(gen):
   golly.error('Sorry, but "' + gen + '" is not a valid integer.')
else:
   previousgen=gen
   goto(gen.replace(",",""))
if os.access(GotoINIFileName, os.W_OK) or not os.access(GotoINIFileName, os.F_OK):
   try:
      f=open(GotoINIFileName, 'w')
      f.write(previousgen)
      f.close()
   except:
      golly.show("Unable to save defaults to " + GotoINIFileName)

File: Scripts/Perl/goto.pl
# Go to a requested generation.  The given generation can be an
# absolute number like 1,000,000 (commas are optional) or a number
# relative to the current generation like +9 or -6.  If the target
# generation is less than the current generation then we go back
# to the starting generation (normally 0) and advance to the target.
# Authors: Andrew Trevorrow and Dave Greene, June 2007.

use strict;
use Time::HiRes qw ( time );

# --------------------------------------------------------------------

sub go_to {
   my $gen = shift;
   my $currgen = g_getgen();
   my $newgen = $gen;
   if (substr($gen,0,1) eq '+') {
      $newgen = $currgen + $gen;
   } elsif (substr($gen,0,1) eq '-') {
      my $n = -$gen;
      if ($currgen > $n) {
         $newgen = $currgen - $n;
      } else {
         $newgen = 0;
      }
   }
   
   if ($newgen < $currgen) {
      # try to go back to starting gen (not necessarily 0) and
      # then forwards to newgen; note that reset() also restores
      # algorithm and/or rule, so too bad if user changed those
      # after the starting info was saved;
      # first save current location and scale
      my ($midx, $midy) = g_getpos();
      my $mag = g_getmag();
      g_reset();
      # restore location and scale
      g_setpos($midx, $midy);
      g_setmag($mag);
      # current gen might be > 0 if user loaded a pattern file
      # that set the gen count
      $currgen = g_getgen();
      if ($newgen < $currgen) {
         g_error("Can't go back any further; pattern was saved ".
                 "at generation $currgen.");
         return;
      }
   }
   
   g_show("Hit escape to abort...");
   my $oldsecs = time;
   while ($currgen < $newgen) {
      if (g_empty()) {
         g_show("Pattern is empty.");
         return;
      }
      # using g_step() as a faster method for getting to newgen
      # is left as an exercise for the reader :)
      g_run(1);
      $currgen += 1;
      g_dokey( g_getkey() );              # allow keyboard interaction
      my $newsecs = time;
      if ($newsecs - $oldsecs >= 1.0) {   # do an update every sec
         $oldsecs = $newsecs;
         g_update();
      }
   }
   g_show("");
}

# --------------------------------------------------------------------

# use same file name as in goto.py
my $GotoINIFileName = g_datadir()."goto.ini";
my $previousgen = "";

if (open(INFILE, $GotoINIFileName)) {
   $previousgen = <INFILE>;
   close INFILE;
}

my $gen = g_getstring("Enter the desired generation number,\n".
                      "or -n/+n to go back/forwards by n:",
                      $previousgen, "Go to generation");
if ($gen eq "") {
   g_exit();
} elsif ($gen eq '+' or $gen eq '-') {
   # clear the default
   $previousgen = "";
} elsif ($gen !~ /^[+-]?\d[\d,]*$/) {
   g_exit("Sorry, but \"$gen\" is not a valid integer.");
} else {
   $previousgen = $gen;
   $gen =~ s/,//g;
   go_to($gen);
}

if (not open(OUTFILE, ">".$GotoINIFileName)) {
   g_warn("Can't save gen in $GotoINIFileName:\n$!");
} else {
   print OUTFILE $previousgen;
   close OUTFILE;
}
